pkg/webhook/validators/channel_validator.go

```diff
@@ -0,0 +1,133 @@
+// in the cluster (https://github.com/knative/eventing/issues/779)
+// To check if Channel Provisioner is installed, and informer needs to be created on ChanneProvisioners and provided to the NewChannelValidator func.
+// ChannelValidatorImpl will then use the Store (Or Indexer) inside this informer to check if the ChannelProvisioner exists or not and will
+// not make a API Server call in the admission webhook. However the Informer will get Add/Delete/Update events on ChannelProvisioner and 
+// the underlying Store/Indexer will be kept up to date.
+type ChannelValidatorImpl struct {
+	store  cache.Store
+	logger *zap.Logger
+}
+
+var _ eventingv1alpha1.ChannelValidator = (*ChannelValidatorImpl)(nil)
+
+
+func NewChannelValidator(informer cache.SharedIndexInformer, logger *zap.Logger) (*ChannelValidatorImpl, error) {
+	if informer == nil {
+		return nil, errors.New("Required argument is nil: informer")
+	}
+	if logger == nil {
+		return nil, errors.New("Required argument is nil: logger")
+	}
+	return &ChannelValidatorImpl{
+		store:  informer.GetStore(),
+		logger: logger.With(zap.String("role", "ChannelValidator")),
+	}, nil
+}
+
+func (cv *ChannelValidatorImpl) Validate(ch *eventingv1alpha1.Channel) *apis.FieldError {
+	errs := cv.validateChannelSpec(&ch.Spec).ViaField("spec")
+	return errs
+}
+
+func (cv *ChannelValidatorImpl) validateChannelSpec(cs *eventingv1alpha1.ChannelSpec) *apis.FieldError {
+	var errs *apis.FieldError
+
+	errs = errs.Also(cv.validateProvisioner(cs.Provisioner))
+	errs = errs.Also(cv.validateSubscribable(cs.Subscribable))
+
+	return errs
+}
+
+func (cv *ChannelValidatorImpl) validateProvisioner(provs *corev1.ObjectReference) *apis.FieldError {
+	if provs == nil {
+		return apis.ErrMissingField("provisioner")
+	}
+	if !cv.supportedProvisioners(provs) {
+		return &apis.FieldError{
+			Message: "Invalid",
+			Details: "This kind of channel provisioner is not supported. Supported provisioners inclue: {apiVersion:eventing.knative.dev/v1alpha1, kind:ClusterChannelProvisioner}",
+			Paths:   []string{"provisioner"},
+		}
+	}
+
+	if !cv.provisionerInstalled(provs) {
+		return &apis.FieldError{
+			Message: "Not found",
+			Details: fmt.Sprintf("Specified channel provisioner 'name:%v, kind:%v, apiVersion:%v ' not installed in cluster.", provs.Name, provs.Kind, provs.APIVersion),
+			Paths:   []string{"provisioner"},
+		}
+	}
+	return nil
+}
+
+func (cv *ChannelValidatorImpl) supportedProvisioners(obj *corev1.ObjectReference) bool {
+	if obj != nil &&
+		obj.APIVersion == "eventing.knative.dev/v1alpha1" &&
+		obj.Kind == "ClusterChannelProvisioner" {
+		return true
+	}
+	return false
+}
+
+func (cv *ChannelValidatorImpl) provisionerInstalled(obj *corev1.ObjectReference) bool {
+	if obj != nil {
+		_, exists, err := cv.store.GetByKey(obj.Name)
+
+		if err != nil {
+			cv.logger.Error("Error while retrieving ChannelProvisioner from the informer store.")
```

Comment:
This breaks late binding. This really should be a status on the channel, not a failure to create the channel.

Reply:
I explored that option and discarded it. The problem with Status is that now we will need a common Controller that controls some of the Channel Status while there already exists Channel-type-specific  controller that controls its Status.
SO we either make it mandatory for each Channel Controller to take care of this which is difficult to maintain and guarantee, or have two controller control the status of a single object which is also a no-no.

So finally after discussing with Adam we decided to stick with a Webhook which will help improve user experience but as you mentioned will break late binding and not catch issues later on if say post Channel provisioning the ChannelProvisioner is deleted.

Thoughts?